### DISCUSSION TITLE
Multiple _GraphQLSchemaProvider_s in one application

### DISCUSSION THREAD

**Discussion by ruanfi** [graphql-java/graphql-java#2198] [Q&A]
Hi.
I am creating a GraphQL in java code dynamically from the proprietary schema description. I got one _GraphQLSchemaProvider_ per tenant. So requests from different tenants will use different _GraphQLSchemaProvider_ object.
The questions are:

1. How heavy _GraphQLSchemaProvider_ is? What if I will have around 100 of it in one application.
2. I get a _GraphQLSchema_ instance from it and provide into _GraphQLSingleInvocationInput_ for my own _GraphQLSpringInvocationInputFactory_ redefined. Is it thread-safe? The whole solution is using WebFlux for reactivness.
3. Are there any other approaches I can take, the main goal is to have dynamic schema per tenant, not highly modifiable, just from time to time with proprietary schema changes inside storage, on the fly.
Thanks.

**Comment by bbakerman** [comment]
> How heavy GraphQLSchemaProvider is? 

The "heavy" ness is really the GraphqlSchema provided by it.  How heavy is a function of how big your schemas definitions are, eg how many types and how many fields.

The schema is a very memory oriented structure - so you will need to run this up, take a memory dump with a tool like YourKit or JProfiler and multiply it out by 100.  This is your memory overhead.

**Comment by dfa1** [comment]
> 1. How heavy _GraphQLSchemaProvider_ is? What if I will have around 100 of it in one application.

In our application we have two graphql instances of a pretty big schema (around 40K lines of SDL).
Last time I checked was about 30MB of heap (according to yourkit). 

> 3. Are there any other approaches I can take, the main goal is to have dynamic schema per tenant, not highly modifiable, just from time to time with proprietary schema changes inside storage, on the fly.

You could try to use `GraphQL#transform`
